File: java/flowcamel-core/src/main/java/com/flowcamel/core/uri/ComponentUri.java
package com.flowcamel.core.uri;

import com.flowcamel.core.catalog.CamelCatalogProperty;
import com.flowcamel.core.catalog.CatalogComponentEntry;
import com.flowcamel.core.model.EndpointDescriptor;
import com.flowcamel.core.properties.ComponentProperties;
import com.flowcamel.core.properties.ComponentProperties.ComponentRole;
import com.flowcamel.core.registry.CatalogRegistry;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class ComponentUri {
  private static final String[] SEPARATORS = {"://", "//", ":", "/", "#"};

  private ComponentUri() {}

  public static List<String> parseSyntax(String syntax) {
    String simplified = syntax;
    for (String s : SEPARATORS) simplified = simplified.replace(s, ":");
    List<String> parts = new ArrayList<>();
    for (String p : simplified.split(":")) if (!p.isEmpty()) parts.add(p);
    return parts;
  }

  public static List<String> getSyntaxSeparators(String syntax) {
    List<String> parts = parseSyntax(syntax);
    List<String> result = new ArrayList<>();
    String str = "";
    for (int index = 0; index < parts.size() - 1; index++) {
      String part = parts.get(index);
      String next = parts.get(index + 1);
      int start = syntax.indexOf(part, str.length()) + part.length();
      int end = syntax.indexOf(next, start);
      result.add(syntax.substring(start, end));
      str = str + part + syntax.substring(start, end);
    }
    return result;
  }

  public static Map<String, String> getUriParts(String uri, String scheme) {
    Map<String, String> result = new LinkedHashMap<>();
    CatalogComponentEntry desc = CatalogRegistry.resolveEntry(scheme).orElse(null);
    if (desc == null || uri == null || uri.isEmpty()) return result;

    String name = desc.component.name;
    if ("salesforce".equals(name)) {
      String[] parts = uri.split(":");
      if (parts.length == 2) {
        result.put("operationName", parts[1]);
        result.put("topicName", "");
      } else if (parts.length == 3) {
        result.put("operationName", parts[1]);
        result.put("topicName", parts[2]);
      }
      return result;
    }
    if ("cxf".equals(name)) {
      String[] cxfParts = uri.split(":");
      if (cxfParts.length == 3 && "bean".equals(cxfParts[1]) && cxfParts[2] != null) {
        result.put("beanId", cxfParts[1] + ":" + cxfParts[2]);
      }
      if (cxfParts.length == 2 && cxfParts[1] != null && cxfParts[1].startsWith("//")) {
        result.put("address", cxfParts[1]);
      }
      return result;
    }
    if ("jt400".equals(name)) {
      String[] jt = uri.replace(".", ":").replace("/", ":").replace("@", ":").split(":");
      if (jt.length > 1) result.put("userID", jt[1]);
      if (jt.length > 2) result.put("password", jt[2]);
      if (jt.length > 3) result.put("systemName", jt[3]);
      if (jt.length > 4) result.put("objectPath", jt[4]);
      if (jt.length > 5) result.put("type", jt[5]);
      return result;
    }

    String syntax = desc.component.syntax;
    List<String> syntaxParts = parseSyntax(syntax);
    List<String> syntaxSeparators = getSyntaxSeparators(syntax);
    String newUri = uri.equals(name) ? name + String.join("", syntaxSeparators) : uri;

    for (int index = 0; index < syntaxParts.size() - 1; index++) {
      if (index == 0) continue;
      String part = syntaxParts.get(index);
      String startSeparator = syntaxSeparators.get(index - 1);
      String endSeparator = index < syntaxSeparators.size() ? syntaxSeparators.get(index) : "";
      int start = newUri.indexOf(startSeparator) + startSeparator.length();
      int end = endSeparator.isEmpty() ? newUri.length() : newUri.indexOf(endSeparator, start);
      result.put(part, newUri.substring(start, end));
      newUri = newUri.substring(end);
    }
    return result;
  }

  public static EndpointDescriptor splitEndpointUri(String fullUri, String scheme) {
    if (fullUri == null || fullUri.isEmpty()) return EndpointDescriptor.of(scheme);
    String base = fullUri.split("\\?")[0];
    Map<String, String> parameters = new LinkedHashMap<>(getUriParts(base, scheme));
    parameters.putAll(parseQueryParams(fullUri));
    String schemeOnly = scheme != null && !scheme.isEmpty() ? scheme : fullUri.split(":")[0].split("\\?")[0];
    boolean hasComplexPath = fullUri.contains("://") || fullUri.contains("@");
    if (parameters.isEmpty() || hasComplexPath) return EndpointDescriptor.of(fullUri);
    return EndpointDescriptor.withParams(schemeOnly, parameters);
  }

  private static Map<String, String> parseQueryParams(String fullUri) {
    int qIdx = fullUri.indexOf('?');
    if (qIdx < 0) return Map.of();
    Map<String, String> out = new LinkedHashMap<>();
    for (String pair : fullUri.substring(qIdx + 1).split("&")) {
      if (pair.isEmpty()) continue;
      int eq = pair.indexOf('=');
      String key = URLDecoder.decode(eq >= 0 ? pair.substring(0, eq) : pair, StandardCharsets.UTF_8);
      String val = URLDecoder.decode(eq >= 0 ? pair.substring(eq + 1) : "", StandardCharsets.UTF_8);
      if (!key.isEmpty()) out.put(key, val);
    }
    return out;
  }

  public static String buildUriFromCatalog(String scheme, ComponentRole role, Map<String, String> props) {
    CatalogComponentEntry desc = CatalogRegistry.resolveEntry(scheme).orElse(null);
    if (desc == null) return "";
    List<String> parts = parseSyntax(desc.component.syntax);
    List<String> separators = getSyntaxSeparators(desc.component.syntax);
    Set<String> pathNames = new HashSet<>(getPathParamNames(scheme));

    StringBuilder uri = new StringBuilder(parts.getFirst());
    for (int i = 1; i < parts.size(); i++) {
      String param = parts.get(i);
      String val = props.get(param);
      if (val == null || val.isEmpty()) return "";
      uri.append(separators.get(i - 1)).append(val);
    }
    return appendQuery(uri.toString(), props, ComponentProperties.getComponentProperties(scheme, role), pathNames);
  }

  private static List<String> getPathParamNames(String scheme) {
    CatalogComponentEntry desc = CatalogRegistry.resolveEntry(scheme).orElse(null);
    if (desc == null) return List.of();
    List<String> fromKind = ComponentProperties.getComponentProperties(scheme, ComponentRole.consumer).stream()
        .filter(p -> "path".equals(p.kind))
        .map(p -> p.name)
        .toList();
    if (!fromKind.isEmpty()) return fromKind;
    List<String> syntaxParts = parseSyntax(desc.component.syntax);
    return syntaxParts.size() > 1 ? syntaxParts.subList(1, syntaxParts.size()) : List.of();
  }

  private static String appendQuery(
      String uri, Map<String, String> props, List<CamelCatalogProperty> catalogProps, Set<String> pathNames) {
    List<String> params = new ArrayList<>();
    for (CamelCatalogProperty p : catalogProps) {
      if (pathNames.contains(p.name) || "path".equals(p.kind)) continue;
      String val = props.get(p.name);
      if (val == null || val.isEmpty()) continue;
      params.add(p.name + "=" + java.net.URLEncoder.encode(val, StandardCharsets.UTF_8));
    }
    if (params.isEmpty()) return uri;
    return uri + (uri.contains("?") ? "&" : "?") + String.join("&", params);
  }
}

File: packages/core/src/api/ComponentUri.ts
import type { CamelCatalogProperty } from '../model/CamelCatalog.js';
import { resolvePropForEmit } from './ConfigRefs.js';
import { getCatalogComponent } from './CatalogRegistry.js';
import { getComponentProperties, type ComponentRole } from './ComponentProperties.js';

const SEPARATORS = ['://', '//', ':', '/', '#'];

export function parseSyntax(syntax: string): string[] {
  let simplified = syntax;
  for (const s of SEPARATORS) simplified = simplified.replaceAll(s, ':');
  return simplified.split(':').filter(Boolean);
}

export function getSyntaxSeparators(syntax: string): string[] {
  const result: string[] = [];
  const parts = parseSyntax(syntax);
  let str = '';
  parts.forEach((part, index) => {
    if (index < parts.length - 1) {
      const next = parts[index + 1];
      if (!next) return;
      const start = syntax.indexOf(part, str.length) + part.length;
      const end = syntax.indexOf(next, start);
      result.push(syntax.substring(start, end));
      str = str + part + syntax.substring(start, end);
    }
  });
  return result;
}

function getPathParamNames(scheme: string): string[] {
  const desc = getCatalogComponent(scheme);
  if (!desc) return [];
  const fromKind = desc.properties.filter((p) => p.kind === 'path').map((p) => p.name);
  if (fromKind.length > 0) return fromKind;
  return parseSyntax(desc.component.syntax).slice(1);
}

export function getUriParts(uri: string, scheme: string): Map<string, string> {
  const result = new Map<string, string>();
  const desc = getCatalogComponent(scheme);
  if (!desc || !uri) return result;

  const name = desc.component.name;

  if (name === 'salesforce') {
    const parts = uri.split(':');
    if (parts.length === 2) {
      result.set('operationName', parts[1] ?? '');
      result.set('topicName', '');
    } else if (parts.length === 3) {
      result.set('operationName', parts[1] ?? '');
      result.set('topicName', parts[2] ?? '');
    }
    return result;
  }

  if (name === 'cxf') {
    const cxfParts = uri.split(':');
    const firstPart = cxfParts[1];
    const secondPart = cxfParts[2];
    if (cxfParts.length === 3 && firstPart === 'bean' && secondPart) {
      result.set('beanId', `${firstPart}:${secondPart}`);
    }
    if (cxfParts.length === 2 && firstPart?.startsWith('//')) {
      result.set('address', firstPart);
    }
    return result;
  }

  if (name === 'jt400') {
    const jt400Parts = uri
      .split('.')
      .join(':')
      .split('/')
      .join(':')
      .split('@')
      .join(':')
      .split(':');
    result.set('userID', jt400Parts[1] ?? '');
    result.set('password', jt400Parts[2] ?? '');
    result.set('systemName', jt400Parts[3] ?? '');
    result.set('objectPath', jt400Parts[4] ?? '');
    result.set('type', jt400Parts[5] ?? '');
    return result;
  }

  const syntax = desc.component.syntax;
  const syntaxParts = parseSyntax(syntax);
  const syntaxSeparators = getSyntaxSeparators(syntax);
  let newUri = uri === name ? `${name}${syntaxSeparators.join('')}` : uri;

  syntaxParts
    .filter((_, i) => i > 0)
    .forEach((part, index) => {
      if (index < syntaxParts.length - 1) {
        const startSeparator = syntaxSeparators[index] ?? '';
        const endSeparator = syntaxSeparators[index + 1];
        const start = newUri.indexOf(startSeparator) + startSeparator.length;
        const end = endSeparator ? newUri.indexOf(endSeparator, start) : newUri.length;
        result.set(part, newUri.substring(start, end));
        newUri = newUri.substring(end);
      }
    });

  return result;
}

export interface EndpointDescriptor {
  uri: string;
  parameters?: Record<string, string>;
}

function parseQueryParams(fullUri: string): Record<string, string> {
  const qIdx = fullUri.indexOf('?');
  if (qIdx < 0) return {};
  const out: Record<string, string> = {};
  for (const pair of fullUri.slice(qIdx + 1).split('&')) {
    if (!pair) continue;
    const eq = pair.indexOf('=');
    const key = decodeURIComponent(eq >= 0 ? pair.slice(0, eq) : pair);
    const val = decodeURIComponent(eq >= 0 ? pair.slice(eq + 1) : '');
    if (key) out[key] = val;
  }
  return out;
}

/** Split a built endpoint URI into scheme + parameters (Karavan-style YAML DSL). */
export function splitEndpointUri(fullUri: string, scheme: string): EndpointDescriptor {
  if (!fullUri) return { uri: scheme };

  const pathParts = getUriParts(fullUri.split('?')[0] ?? fullUri, scheme);
  const parameters: Record<string, string> = {};
  for (const [k, v] of pathParts) {
    if (v !== undefined && v !== '') parameters[k] = v;
  }
  Object.assign(parameters, parseQueryParams(fullUri));

  const schemeOnly = scheme || fullUri.split(':')[0]?.split('?')[0] || fullUri;
  const hasComplexPath = fullUri.includes('://') || fullUri.includes('@');

  if (Object.keys(parameters).length === 0 || hasComplexPath) {
    return { uri: fullUri };
  }

  return { uri: schemeOnly, parameters };
}

function appendQuery(
  uri: string,
  props: Record<string, string>,
  catalogProps: CamelCatalogProperty[],
  pathNames: Set<string>
): string {
  const params: string[] = [];
  for (const p of catalogProps) {
    if (pathNames.has(p.name) || p.kind === 'path') continue;
    const val = resolvePropForEmit(props[p.name]);
    if (!val) continue;
    params.push(`${p.name}=${encodeURIComponent(val)}`);
  }
  if (params.length === 0) return uri;
  return uri + (uri.includes('?') ? '&' : '?') + params.join('&');
}

export function buildUriFromCatalog(
  scheme: string,
  role: ComponentRole,
  props: Record<string, string>
): string {
  const desc = getCatalogComponent(scheme);
  if (!desc) return '';

  const parts = parseSyntax(desc.component.syntax);
  const separators = getSyntaxSeparators(desc.component.syntax);
  const pathNames = new Set(getPathParamNames(scheme));

  let uri = parts[0] ?? scheme;
  for (let i = 1; i < parts.length; i++) {
    const param = parts[i];
    if (!param) return '';
    const val = resolvePropForEmit(props[param]);
    if (!val) return '';
    uri += (separators[i - 1] ?? ':') + val;
  }

  return appendQuery(uri, props, getComponentProperties(scheme, role), pathNames);
}
